src/core/border_router/routing_manager.cpp

```diff
@@ -550,26 +550,22 @@ void RoutingManager::UnpublishExternalRoute(const Ip6::Prefix &aPrefix)
 
 void RoutingManager::EvaluateOnLinkPrefix(void)
 {
-    const Ip6::Prefix *smallestOnLinkPrefix = nullptr;
+    const ExternalPrefix *onLinkPrefix = nullptr;
 
     // We don't evaluate on-link prefix if we are doing Router Solicitation.
     VerifyOrExit(!IsRouterSolicitationInProgress());
 
     for (const ExternalPrefix &prefix : mDiscoveredPrefixes)
     {
-        if (!prefix.IsOnLinkPrefix() || prefix.IsDeprecated())
+        if (prefix.IsOnLinkPrefix() && !prefix.IsDeprecated())
```

Comment:
I think we still need to select the smallest on-link prefix. 

There are still cases where it's necessary:
- When the AP router advertises a global GUA prefix
- When a BR of another Thread network advertises an on-link prefix
- When a BR using old implementation that advertises random ULA on-link prefix exists in the network

Thoughts? @wgtdkp

Reply:
+1 on simon's good summary. We better not depend on the fact that all BRs in the same Thread network will use the same on-link prefix, it doesn't give the guarantee of receiving the same on-link prefixes on infra link.

Reply:
The fact that `smallestOnLinkPrefix` is smallest is used under one code path, where it is not `nullptr` and we are already advertising local on-link prefix,`mIsAdvertisingLocalOnLinkPrefix == true` and we want to decide if to stop advertising or not.

Let me try to go trough each case and check the difference in behavior between current code and the changes in this PR

#### (A) If AP router is advertising a (global GUA) prefix
We already have a an on-link on infra side, so we don't need to advertise `mLocalOnLinkPrefix`.
- Current code: We may still continue to advertise `mLocalOnLinkPrefix` is it is smaller than AP's prefix and we end up with two on-link prefixes (we have a comment about this in the code). I guess it is fine to have two prefixes, but it is not not ideal.
- New code: AP's prefix wins and we don't advertise `mLocalOnLinkPrefix`

#### (B) When a BR of another Thread network advertises an on-link prefix
From our perspective (on the BR deciding whether to advertise or not), this is like the first situation (AP router advertising). We have an acceptable on-link and don't need to advertise `mLocalOnLinkPrefix`.
- Current code: We will stop advertising only if our prefix is larger one. Otherwise we continue advertising (hoping the other BRs to to back off and stop).
- New PR: We deprecate and stop advertising.

One situation I can see if multiple BRs (from different thread networks) try to compete (add their own prefix and then back off when they see the other BR's prefix). We can rely on random RA tx times (jitters) to have one win. Or we can keep the "smallest prefix win" logic.

#### (C) When a BR using old implementation that advertises random ULA on-link prefix exists in the network**
This looks and behaves similar to the previous case. 
And I think this would be atemporary situation as BRs should use the new model (prefix derived from xpanid soon). Not sure if we require as 1.3 cert test or not?

Reply:
Thinking more on this, I think it may be good to continue using the "smallest prefix" for case (B) above (instead of relying on random jitter times between RA tx).

Reply:
Regarding (**B**):
> One situation I can see: if multiple BRs (from different thread networks) try to compete (add their own prefix and then back off when they see the other BR's prefix). We can rely on random RA tx times (jitters) to have one win. Or we can keep the "smallest prefix win" logic.

This is actually the initial motivation why we added the logic of selecting the smallest on-link prefix. In case multiple BRs advertise prefixes at the same time, we can let all BRs to withdraw their prefixes and retry with a random delay, but it would be better to have deterministic strategy to decide the prefix which shouldn't be withdrawn at all (faster on-link prefix establishment).

Reply:
@abtink 
Thanks for the case-by-case analysis. I would agree that there are pros and cons to both strategies, and there is no clear winner. 

I think selecting the smallest on-link prefix works better overall. For case (A), the global GUA prefix advertised by the AP router (usually starts with 20) is usually smaller than the ULA prefix (starts with fd).

---

Thread spec 9.6.2 refers to section 3.1.3 of draft-lemon-stub-networks-02 for more details. We can find the latest version at https://datatracker.ietf.org/doc/html/draft-lemon-stub-networks-03#section-3.1.3.

The process defined in [draft-lemon-stub-networks-03#section-3.1.3](https://datatracker.ietf.org/doc/html/draft-lemon-stub-networks-03#section-3.1.3) requires a BR  to always deprecate its own on-link prefix when another preferred on-link prefix is detected, same as this PR, but also provide a mechanism for BR to re-advertise its on-link prefix by comparing valid lifetimes. 

Whether [draft-lemon-stub-networks-03#section-3.1.3](https://datatracker.ietf.org/doc/html/draft-lemon-stub-networks-03#section-3.1.3) is optimal is up to debate, but I think we can either fully implement it, or keep the current strategy (selecting the smallest on-link prefix) as is. 

Thoughts? @wgtdkp @abtink

Reply:
BTW, a BR that selects the smallest on-link prefix can co-work other BRs that implement [draft-lemon-stub-networks-03#section-3.1.3](https://datatracker.ietf.org/doc/html/draft-lemon-stub-networks-03#section-3.1.3).

Reply:
Thanks guys! This is nice discussion.
I am going to close this PR. Let's keep the smallest prefix wins logic (I agree that it will very useful specially for (B) and cab help with other situations as well).

Regarding https://datatracker.ietf.org/doc/html/draft-lemon-stub-networks-03#section-3.1.3, I did not know. Thanks. My initial thought on this is that I can see that it can work, but I feel it may be somewhat complex logic for addressing a corner case.

Reply:
> Regarding https://datatracker.ietf.org/doc/html/draft-lemon-stub-networks-03#section-3.1.3, I did not know. Thanks. My initial thought on this is that I can see that it can work, but I feel it may be somewhat complex logic for addressing a corner case.

I think the benefits of https://datatracker.ietf.org/doc/html/draft-lemon-stub-networks-03#section-3.1.3 is that it can always converge to one on-link prefix. But agree that it's much more complex than selecting the smallest.